databuilder/query_engines/sqlite.py

```diff
@@ -123,14 +156,112 @@ def get_sql_select_column(self, node):
         source = self.get_sql(node.source)
         return source.c[node.name]
 
-    # We have to apply caching here otherwise we generate disinct objects representing
+    # We have to apply caching here otherwise we generate distinct objects representing
     # the same table and this confuses SQLAlchemy into generating queries with ambiguous
     # table references
+    @get_sql.register(SelectTable)
     @get_sql.register(SelectPatientTable)
     @cache
     def get_sql_select_table(self, node):
         return self.backend.get_table_expression(node.name)
 
+    # We ignore Filter and Sort operations completely at this point in the code and just
+    # pass the underlying table reference through. It's only later, when building the
+    # SELECT query for a given Frame, that we make use of these. This is in order to
+    # mirror the semantics of SQL whereby columns are selected directly from the
+    # underlying table and filters and sorts are handled separately using WHERE/ORDER BY
+    # clauses.
```

Comment:
I'm going to have to take your and Hypothesis' word for it that this works. I can't reason this through sufficiently to convince myself that it's always okay to lift these up to the outside without leaving any trace.

Reply:
Yeah, this is all a bit tricky to reason about. I'd say this move was the key thing that unstuck me in writing this code though. It might be easier to talk about it and see if we can come up with a more convincing narrative that we can introduce here.

Reply:
I wonder if this is another candidate for (imaginary future) pre-translation transformation.

Reply:
Yes, I've wondered exactly that too.